### PR TITLE
[15721] Index Suggestion

### PR DESCRIPTION
**Pull request by sivaprasadsudhir**
This PR implements a search based algorithm for suggesting the best set of indexes for a given workload based on the AutoAdmin paper from Microsoft. 

This involves 
- WhatIfAPI - that returns the cost of a query for a hypothetical index configuration (brain/what_if_index, optimizer/optimizer)

- IndexSelection - Runs the search based algorithm. This involves 3 steps in each iteration - candidate index selection to choose indexes that benefit at least one query in the workload, configuration enumeration to explore the search space of the indexes and choose the top k indexes for the best configuration, and multi-column index generation to generate the indexes for the next iteration. (brain/index_selection*)

### CODE REVIEW

File: src/brain/index_selection.cpp [hunk 1]
```diff
@@ -0,0 +1,449 @@
...
+  // The following rules define what indexable columns are:
+  // 1. A column that appears in the WHERE clause with format
+  //    ==> Column OP Expr <==
+  //    OP such as {=, <, >, <=, >=, LIKE, etc.}
+  //    Column is a table column name.
+  // 2. GROUP BY (if present)
+  // 3. ORDER BY (if present)
+  // 4. all updated columns for UPDATE query.
+
+  union {
```
Comment (tli2) [inline src/brain/index_selection.cpp]: Using the union here is a bit weird since you can just declare local variables in each of the switch cases. The compiler should be able to recognize that life time don't overlap and do this automatically. Using a somewhat esoteric feature like the union tends to confuse people.

File: src/brain/index_selection.cpp [hunk 2]
```diff
@@ -0,0 +1,449 @@
...
+    return;
+  }
+  auto expr_type = where_expr->GetExpressionType();
+  const expression::AbstractExpression *left_child;
+  const expression::AbstractExpression *right_child;
+  const expression::TupleValueExpression *tuple_child;
+
+  switch (expr_type) {
+    case ExpressionType::COMPARE_EQUAL:
+      PELOTON_FALLTHROUGH;
```
Comment (tli2) [inline src/brain/index_selection.cpp]: Nit: I don't know if we even enable fallthrough checks in our build, but it seems like in other cases we only use this compiler keyword when there is a body of code after the tag (i.e. case ...: ...; PELOTON_FALLTHROUGH case:...)

I am relatively certain that if the body is empty people just leave out the fallthrough?

File: src/include/catalog/index_catalog.h [hunk 3]
```diff
@@ -6,33 +6,13 @@
 //
 // Identification: src/include/catalog/index_catalog.h
 //
-// Copyright (c) 2015-17, Carnegie Mellon University Index Group
```
Comment (tli2) [inline src/include/catalog/index_catalog.h]: Why is this removed?

File: src/include/brain/what_if_index.h [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+#include "catalog/column_catalog.h"
+#include "catalog/database_catalog.h"
+#include "catalog/index_catalog.h"
+#include "catalog/table_catalog.h"
+#include "common/internal_types.h"
+#include "optimizer/optimizer.h"
+#include "parser/postgresparser.h"
+
+namespace parser {
+  class SQLStatementList;
```
Comment (tli2) [inline src/include/brain/what_if_index.h]: nit: some of these look unused.

File: src/brain/index_selection.cpp [hunk 5]
```diff
@@ -0,0 +1,449 @@
...
+  }
+}
+
+void IndexSelection::ExhaustiveEnumeration(IndexConfiguration &indexes,
+                                           IndexConfiguration &top_indexes,
+                                           Workload &workload) {
+  // Get the best m index configurations using the naive enumeration algorithm
+  // The naive algorithm gets all the possible subsets of size <= m and then
+  // returns the cheapest m indexes
+  assert(context_.naive_enumeration_threshold_ <= indexes.GetIndexCount());
```
Comment (tli2) [inline src/brain/index_selection.cpp]: nit: use peloton assert

File: src/brain/index_selection.cpp [hunk 6]
```diff
@@ -0,0 +1,449 @@
...
+
+  // Define a set ordering of (index config, cost) and define the ordering in
+  // the set
+  std::set<std::pair<IndexConfiguration, double>, IndexConfigComparator>
+      running_index_config(workload);
```
Comment (tli2) [inline src/brain/index_selection.cpp]: nit: declare those together to avoid writing the long typename over and over again

File: src/include/brain/index_selection.h [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+   * @param candidate_config - new candidate index to be pruned.
+   * @param admissible_config - admissible index set of the queries
+   * @param workload - queries
+   */
+  void GenerateCandidateIndexes(IndexConfiguration &candidate_config,
+                                IndexConfiguration &admissible_config,
+                                Workload &workload);
+
+  /**
+   * @brief gets the top k cheapest indexes for the workload
```
Comment (tli2) [inline src/include/brain/index_selection.h]: cheapest is a bit confusing here because that seems to imply the cost of building the index

File: src/include/brain/index_selection_util.h [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+    auto parser = parser::PostgresParser::GetInstance();
+    auto txn = txn_manager.BeginTransaction();
+
+    std::unique_ptr<binder::BindNodeVisitor> binder(
+      new binder::BindNodeVisitor(txn, database_name));
+
+    // Parse and bind every query. Store the results in the workload vector.
+    for (auto it = queries.begin(); it != queries.end(); it++) {
+      auto query = *it;
+      LOG_INFO("Query: %s", query.c_str());
```
Comment (tli2) [inline src/include/brain/index_selection_util.h]: nit: maybe change this to debug or trace

File: src/brain/what_if_index.cpp [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+  LOG_DEBUG("Got cost %lf", opt_info_obj->cost);
+
+  txn_manager.CommitTransaction(txn);
+  return opt_info_obj;
+}
+
+void WhatIfIndex::GetTablesReferenced(parser::SQLStatement *query,
+                                std::vector<std::string> &table_names) {
+  // Only support the DML statements.
+  union {
```
Comment (tli2) [inline src/brain/what_if_index.cpp]: Same as before

File: src/brain/index_selection.cpp [hunk 10]
```diff
@@ -0,0 +1,449 @@
...
+  double cost = 0.0;
+  auto queries = workload.GetQueries();
+  for (auto query : queries) {
+    std::pair<IndexConfiguration, parser::SQLStatement *> state = {config,
+                                                                   query};
+    if (context_.memo_.find(state) != context_.memo_.end()) {
+      cost += context_.memo_[state];
+    } else {
+      auto result =
+          WhatIfIndex::GetCostAndBestPlanTree(query, config, DEFAULT_DB_NAME);
```
Comment (tli2) [inline src/brain/index_selection.cpp]: We are already starting to support multiple databases so we can't hard code a database name here.

File: src/catalog/index_catalog.cpp [hunk 11]
```diff
@@ -51,6 +51,20 @@ IndexCatalogObject::IndexCatalogObject(executor::LogicalTile *tile, int tupleId)
   LOG_TRACE("the size for indexed key is %lu", key_attrs.size());
 }
 
+IndexCatalogObject::IndexCatalogObject(oid_t index_oid, std::string index_name,
+                                       oid_t table_oid, IndexType index_type,
+                                       IndexConstraintType index_constraint,
+                                       bool unique_keys,
+                                       std::set<oid_t> key_attrs) {
+  this->index_oid = index_oid;
```
[the remark this answers is not in the document]
Comment (vkonagar) [reply src/catalog/index_catalog.cpp]: Done

File: src/brain/index_selection.cpp [hunk 12]
```diff
@@ -0,0 +1,449 @@
...
+namespace peloton {
+namespace brain {
+
+IndexSelection::IndexSelection(Workload &query_set, size_t max_index_cols,
+                               size_t enum_threshold, size_t num_indexes)
+    : query_set_(query_set),
+      context_(max_index_cols, enum_threshold, num_indexes) {}
+
+void IndexSelection::GetBestIndexes(IndexConfiguration &final_indexes) {
+  // Figure 4 of the "Index Selection Tool" paper.
```
[the remark this answers is not in the document]
Comment (vkonagar) [reply src/brain/index_selection.cpp]: Done

File: test/brain/index_selection_test.cpp [hunk 13]
```diff
@@ -0,0 +1,330 @@
...
+  LOG_DEBUG("Admissible Indexes: %s", admissible_config.ToString().c_str());
+  LOG_DEBUG("Candidate Indexes: %s", candidate_config.ToString().c_str());
+
+  EXPECT_EQ(admissible_config.GetIndexCount(), 2);
+  // TODO: There is no data in the table. Indexes should not help. Should return 0.
+  // EXPECT_EQ(candidate_config.GetIndexCount(), 0);
+  EXPECT_EQ(candidate_config.GetIndexCount(), 2);
+
+  // Insert some tuples into the table.
+  InsertIntoTable(table_name, 2000);
```
[the remark this answers is not in the document]
Comment (vkonagar) [reply test/brain/index_selection_test.cpp]: Fixed

File: test/brain/index_selection_test.cpp [hunk 14]
```diff
@@ -0,0 +1,330 @@
...
+//===--------------------------------------------------------------------===//
+// IndexSelectionTest
+//===--------------------------------------------------------------------===//
+
+class IndexSelectionTest : public PelotonTest {
+ public:
+  IndexSelectionTest() {}
+
+  // Create a new database
+  void CreateDatabase(std::string db_name) {
```
[the remark this answers is not in the document]
Comment (vkonagar) [reply test/brain/index_selection_test.cpp]: Done

File: src/brain/index_selection.cpp (hunk 1 above)
Comment (vkonagar) [reply src/brain/index_selection.cpp]: Fixed

File: src/include/brain/what_if_index.h (hunk 4 above)
Comment (vkonagar) [reply src/include/brain/what_if_index.h]: Done

File: src/brain/index_selection.cpp (hunk 5 above)
Comment (vkonagar) [reply src/brain/index_selection.cpp]: Done

File: src/brain/index_selection.cpp (hunk 6 above)
Comment (vkonagar) [reply src/brain/index_selection.cpp]: Done

File: src/include/brain/index_selection_util.h (hunk 8 above)
Comment (vkonagar) [reply src/include/brain/index_selection_util.h]: Done

File: src/brain/what_if_index.cpp (hunk 9 above)
Comment (vkonagar) [reply src/brain/what_if_index.cpp]: Done

File: src/brain/index_selection.cpp (hunk 10 above)
Comment (vkonagar) [reply src/brain/index_selection.cpp]: Done

File: test/brain/what_if_index_test.cpp [hunk 15]
```diff
@@ -0,0 +1,332 @@
...
+    auto obj_ptr = new brain::IndexObject(database_oid, table_oid, cols);
+    auto index_obj = std::shared_ptr<brain::IndexObject>(obj_ptr);
+
+    txn_manager.CommitTransaction(txn);
+    return index_obj;
+  }
+};
+
+TEST_F(WhatIfIndexTests, SingleColTest) {
+  std::string table_name = "dummy_table_whatif";
```
Comment (saatviks) [inline test/brain/what_if_index_test.cpp]: Might be useful to put a light test description for context

File: test/brain/what_if_index_test.cpp [hunk 16]
```diff
@@ -0,0 +1,332 @@
...
+  auto parser = parser::PostgresParser::GetInstance();
+  auto txn = txn_manager.BeginTransaction();
+
+  std::unique_ptr<binder::BindNodeVisitor> binder(
+      new binder::BindNodeVisitor(txn, DEFAULT_DB_NAME));
+
+  // Get the first statement.
+  auto sql_statement = stmt_list.get()->GetStatement(0);
+
+  binder->BindNameToNode(sql_statement);
```
Comment (saatviks) [inline test/brain/what_if_index_test.cpp]: Just curious about why this is needed?

File: test/brain/what_if_index_test.cpp [hunk 17]
```diff
@@ -0,0 +1,332 @@
...
+
+  // 1. Get the optimized plan tree without the indexes (sequential scan)
+  auto result = brain::WhatIfIndex::GetCostAndBestPlanTree(
+      sql_statement, config, DEFAULT_DB_NAME);
+  auto cost_without_index = result->cost;
+  LOG_DEBUG("Cost of the query without indexes: %lf", cost_without_index);
+  EXPECT_NE(result->plan, nullptr);
+  LOG_INFO("%s", result->plan->GetInfo().c_str());
+
+  // 2. Get the optimized plan tree with 1 hypothetical indexes (indexes)
```
Comment (saatviks) [inline test/brain/what_if_index_test.cpp]: Mention column names of indexes being checked in a comment

File: test/brain/what_if_index_test.cpp [hunk 18]
```diff
@@ -0,0 +1,332 @@
...
+  LOG_DEBUG("Cost of the query with index: %lf", cost_with_index_1);
+  EXPECT_EQ(cost_without_index, cost_with_index_2);
+
+  config.Clear();
+  config.AddIndexObject(CreateHypotheticalIndex(table_name, {1, 2}));
+  result = brain::WhatIfIndex::GetCostAndBestPlanTree(sql_statement, config,
+                                                      DEFAULT_DB_NAME);
+  auto cost_with_index_3 = result->cost;
+  LOG_DEBUG("Cost of the query with index: %lf", cost_with_index_1);
+  EXPECT_GT(cost_without_index, cost_with_index_3);
```
Comment (saatviks) [inline test/brain/what_if_index_test.cpp]: Test that cost of Index on (b,c) should be better than (a, b) or (a, c)?

File: test/brain/what_if_index_test.cpp [hunk 19]
```diff
@@ -0,0 +1,332 @@
...
+  txn_manager.CommitTransaction(txn);
+
+  // Get the optimized plan tree without the indexes (sequential scan)
+  auto result = brain::WhatIfIndex::GetCostAndBestPlanTree(
+      sql_statement, config, DEFAULT_DB_NAME);
+  auto cost_without_index = result->cost;
+  LOG_DEBUG("Cost of the query without indexes: %lf", cost_without_index);
+
+  // Insert hypothetical catalog objects
+  // Index on cols a, c.
```
Comment (saatviks) [inline test/brain/what_if_index_test.cpp]: Comment fix needed here?

File: test/brain/index_selection_test.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (saatviks) [reply test/brain/index_selection_test.cpp]: It might make sense to later setup a Index selection testing utility - a set of simple and more complex workloads with expected indexes that should be selected.

File: test/brain/index_selection_test.cpp [hunk 20]
```diff
@@ -0,0 +1,377 @@
...
+  GenerateTableStats();
+
+  brain::IndexConfiguration best_config;
+  brain::IndexSelection is(workload, max_index_cols, enumeration_threshold,
+                           num_indexes);
+  is.GetBestIndexes(best_config);
+
+  LOG_INFO("Best Indexes: %s", best_config.ToString().c_str());
+  LOG_INFO("Best Index Count: %ld", best_config.GetIndexCount());
+  EXPECT_EQ(best_config.GetIndexCount(), 5);
```
Comment (saatviks) [inline test/brain/index_selection_test.cpp]: 4 right? (A, B, AB, BC) from above?

File: src/brain/index_selection.cpp [hunk 21]
```diff
@@ -0,0 +1,465 @@
...
+      IndexColsParseGroupByHelper(sql_statement.select_stmt->group_by, indexes);
+      break;
+
+    default:
+      LOG_ERROR("Cannot handle DDL statements");
+      PELOTON_ASSERT(false);
+  }
+}
+
+void IndexSelection::IndexColsParseWhereHelper(
```
Comment (Blade-Lee) [inline src/brain/index_selection.cpp]: Suggestion: why not consider using `GetTupleValueExprs()` function?

File: test/brain/index_selection_test.cpp [hunk 22]
```diff
@@ -0,0 +1,377 @@
...
+  auto queries = workload.GetQueries();
+  for (unsigned long i = 0; i < queries.size(); i++) {
+    brain::Workload w(queries[i], workload.GetDatabaseName());
+    brain::IndexSelection is(w, max_cols, enumeration_threshold, num_indexes);
+
+    brain::IndexConfiguration ic;
+    is.GetAdmissibleIndexes(queries[i], ic);
+    LOG_DEBUG("Admissible indexes %ld, %s", i, ic.ToString().c_str());
+
+    auto indexes = ic.GetIndexes();
```
Comment (Blade-Lee) [inline test/brain/index_selection_test.cpp]: `auto indexes` not used?

File: src/brain/index_selection_util.cpp [hunk 23]
```diff
@@ -0,0 +1,146 @@
...
+}
+
+void IndexConfiguration::AddIndexObject(
+    std::shared_ptr<IndexObject> index_info) {
+  indexes_.insert(index_info);
+}
+
+size_t IndexConfiguration::GetIndexCount() const { return indexes_.size(); }
+
+bool IndexConfiguration::IsEmpty() const { return indexes_.size() == 0; }
```
Comment (saatviks) [inline src/brain/index_selection_util.cpp]: nit: use `.empty()`

File: src/brain/index_selection.cpp [hunk 24]
```diff
@@ -0,0 +1,465 @@
...
+  double global_min_cost = ComputeCost(indexes, workload);
+  double cur_min_cost = global_min_cost;
+  double cur_cost;
+  std::shared_ptr<IndexObject> best_index;
+
+  // go through till you get top k indexes
+  while (current_index_count < k) {
+    // this is the set S so far
+    auto original_indexes = indexes;
+    for (auto index : remaining_indexes.GetIndexes()) {
```
Comment (saatviks) [inline src/brain/index_selection.cpp]: I think this can be a const ref: `auto const & index:`.

File: src/brain/index_selection.cpp [hunk 25]
```diff
@@ -0,0 +1,465 @@
...
+
+  IndexConfiguration new_element;
+
+  // Add an empty configuration as initialization
+  IndexConfiguration empty;
+  // The running index configuration contains the possible subsets generated so
+  // far. It is updated after every iteration
+  running_index_config.insert({empty, 0.0});
+
+  for (auto index : indexes.GetIndexes()) {
```
Comment (saatviks) [inline src/brain/index_selection.cpp]: const ref same as above
